### DISCUSSION TITLE
Dynamic Metafields on Checkout Extensions

### DISCUSSION THREAD

**Discussion by gil--** [Shopify/function-examples#275] [API Ideas]
Would love to see a way to define a Function with a metafield that could then dynamically load customer/product metafields in checkout extensions. Right now, they have to be hard coded within the extension config (.toml) but this feels like something a Shopify Function could improve and make more dynamic for public apps in particular.

**Comment by lemonmade** [comment]
Hey Gil, could you expand on this one a bit? What kind of logic would you be performing in the function to determine what metafields to load that you can't achieve by statically declaring the metafields your UI extensions need up front?

Using the [storefront API](https://shopify.dev/docs/api/checkout-ui-extensions/unstable/apis/standardapi#example-storefront-api-access) from a checkout extension is another way this could be done — is there a reason you don't want to go with that approach?

**Reply by gil--** [reply]
From what I can tell, there's no way to access customer metafields using the storefront api
<img width="463" alt="CleanShot 2023-07-05 at 12 10 44@2x" src="https://github.com/Shopify/function-examples/assets/3484527/a8e06cb5-54df-4cbb-bc9a-ed000c94f643">

A pretty common example use-case is leveraging a customer metafield to populate a custom field in checkout such as a tax id, birthdate, or some other value. Merchants often have their own namespace/key defined for these customer metafield so I can't hardcode them in the .toml config like a private app would. 

The logic would be something akin to a custom Function input.graphql where the data for the query is passed into the extension api. I've always wondered if that's even feasible given the number of extensions loading in checkout would probably cause a bottleneck on performance but maybe if you had something like 5 or even 1 function per checkout where you could define custom data added to the overall extension api.
